src/Microsoft.AspNet.OData.Shared/Formatter/EdmLibHelpers.cs

```diff
@@ -843,6 +842,55 @@ public static bool IsNullable(Type type)
             return !type.IsValueType || Nullable.GetUnderlyingType(type) != null;
         }
 
+        /// <summary>
+        /// Get the expected payload type of an OData path.
+        /// </summary>
+        /// <param name="type">The Type to use.</param>
+        /// <param name="path">The path to use.</param>
+        /// <param name="model">The EdmModel to use.</param>
+        /// <returns>The expected payload type of an OData path.</returns>
+        internal static IEdmTypeReference GetExpectedPayloadType(Type type, ODataPath path, IEdmModel model)
+        {
+            IEdmTypeReference expectedPayloadType = null;
+
+            if (typeof(IEdmObject).IsAssignableFrom(type))
+            {
+                // typeless mode. figure out the expected payload type from the OData Path.
+                IEdmType edmType = path.EdmType;
+                if (edmType != null)
+                {
+                    expectedPayloadType = EdmLibHelpers.ToEdmTypeReference(edmType, isNullable: false);
+                    if (expectedPayloadType.TypeKind() == EdmTypeKind.Collection)
+                    {
+                        IEdmTypeReference elementType = expectedPayloadType.AsCollection().ElementType();
+                        if (elementType.IsEntity())
+                        {
+                            // collection of entities cannot be CREATE/UPDATEd. Instead, the request would contain a single entry.
+                            expectedPayloadType = elementType;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                TryGetInnerTypeForDelta(ref type);
+                expectedPayloadType = model.GetEdmTypeReference(type);
+            }
+
+            return expectedPayloadType;
+        }
+
+        internal static bool TryGetInnerTypeForDelta(ref Type type)
```

Comment:
>internal static bool TryGetInnerTypeForDelta(ref Type type) [](start = 8, length = 59)

why no comments on this function?

Reply:
It moved from a location where it had none but that should be fixed.